### PR TITLE
Improve support and testing for G4 with vecgeom-based solids

### PR DESCRIPTION
**Pull request by sethrj**
- Add g4+vecgeom CI job
- Add orange with assertions CI job
- Configure g4 variables into celeritas
- Add vecgeom to build output when using usolids, and add test of config macros
- Use new macros in geotests

### CODE REVIEW

**Comment by sethrj** [comment]
@esseivaju If you have a minute, could you see if this matches your expectations? I think this was a change you made (but I don't know what version of G4 you were using)
```
  [ RUN      ] FourLevelsTest.small_steps
  /home/runner/work/celeritas/celeritas/test/geocel/CheckedGeoTrackView.cc:302: warning: Finding next step up to 1e-19 [mm] when previous step 50 [mm] was already calculated
  /home/runner/work/celeritas/celeritas/src/geocel/g4/GeantGeoTrackView.hh:478: warning: Boundary direction was inconsistent
  /home/runner/work/celeritas/celeritas/test/geocel/CheckedGeoTrackView.cc:335: warning: Returning zero distance should be prohibited: at {15.99999999999, 10, 10} [cm] along {1, 0, 0}, in World_PV/env1/Shape1
  /home/runner/work/celeritas/celeritas/test/geocel/CheckedGeoTrackView.cc:335: warning: Returning zero distance should be prohibited: at {15.99999999999, 10, 10} [cm] along {1, 0, 0}, in World_PV/env1/Shape1
  unknown file: Failure
  C++ exception with description "/home/runner/work/celeritas/celeritas/src/geocel/g4/GeantGeoTrackView.hh:539:
  celeritas: precondition failed: this->has_next_step()" thrown in the test body.
  
  [  FAILED  ] FourLevelsTest.small_steps (0 ms)
```

**Comment by sethrj** [comment]
@esseivaju This is ready! :)

File: src/corecel/io/BuildOutput.cc [hunk 1]
```diff
@@ -101,7 +101,8 @@ void BuildOutput::output(JsonPimpl* j) const
             CO_ADD_COND_VERS(LARSOFT, LArSoft, larsoft);
             CO_ADD_COND_VERS(ROOT, ROOT, root);
             CO_ADD_COND_VERS(VECGEOM, G4VG, g4vg);
-            CO_ADD_COND_VERS(VECGEOM, VecGeom, vecgeom);
+            CO_ADD_COND_VERS(
+                VECGEOM || CELERITAS_GEANT4_USOLIDS, VecGeom, vecgeom);
```
Comment (esseivaju) [inline src/corecel/io/BuildOutput.cc]: Looks like it works but the way it's expended is a bit confusing 😅

File: src/geocel/GeoTrackInterface.hh [hunk 2]
```diff
@@ -210,7 +210,10 @@ class GeoTrackInterface
      * small distances.
      *
      * \pre \c geo_status() is not \c GeoStatus::boundary_inc .
-     * \post The returned distance is in the range \c (0, max_step] .
+     * \post The returned distance is in the range \c [0, max_step] .
+     * A zero distance must indicate a boundary hit, e.g. within the geometry's
+     * surface tolerance. Call move_to_boundary followed by cross_boundary to
+     * resolve the hit even when no physical displacement is needed.
```
Comment (esseivaju) [inline src/geocel/GeoTrackInterface.hh]: I think the \pre is no longer needed (at least for the GeantGeoTrackView)

File: src/geocel/g4/GeantGeoTrackView.hh [hunk 3]
```diff
@@ -538,10 +534,19 @@ void GeantGeoTrackView::move_to_boundary()
...
     CELER_EXPECT(this->has_next_step());
 
+    if (this->geo_status() == GeoStatus::boundary_inc)
+    {
+        // Already on the surface, e.g. after reversing the direction.
+        CELER_ASSERT(this->is_on_boundary());
+        CELER_ASSERT(next_step_ == 0);
+        next_step_ = -1;
+        return;
+    }
```
Comment (esseivaju) [inline src/geocel/g4/GeantGeoTrackView.hh]: Should we add a test case (and update interface doc) now that move_to_boundary on a reentrant boundary is supported?

**Review by esseivaju — APPROVED** [review]

File: src/corecel/io/BuildOutput.cc (hunk 1 above)
Comment (sethrj) [reply src/corecel/io/BuildOutput.cc]: Yeah I thought I could get away with it ;) I'll clean it up

File: src/geocel/GeoTrackInterface.hh (hunk 2 above)
Comment (sethrj) [reply src/geocel/GeoTrackInterface.hh]: True. I see there's a todo for the ORANGE version of this:
```
\todo Prohibit when GeoStatus::boundary_inc
```

if we know a priori that we're headed into the boundary, then we should avoid the extra find call outside of it. But the boundary state situation is complicated right now… I'll update the docs to note we might want to move this direciton.

File: src/geocel/g4/GeantGeoTrackView.hh (hunk 3 above)
Comment (sethrj) [reply src/geocel/g4/GeantGeoTrackView.hh]: I thought that's what your gap test was supposed to do? (Though I don't see this code hit by the g4-native solid case.)

Comment (sethrj) [reply src/geocel/g4/GeantGeoTrackView.hh]: Let's leave that complexity for another follow-on, and I'll just add a note for now rather than drag this out.

Comment (esseivaju) [reply src/geocel/g4/GeantGeoTrackView.hh]: I was thinking of a different scenario where it's now legal to call  the sequence `find_next_step` --> `move_to_boundary` on an incident boundary, but this is not tested. Though I agree this can be followed-up in another PR.
